Put inference parameters in a dedicated object

maleadt opened the pull request:
Part of #18338, based on #18396 for now because it uses kwargs in inference.

Entry-point function parameters to type inference (`needtree`, `optimize`, `cached`) are now saved in the `InferenceState`, which makes them available in more places. However, using `cached` that way [breaks `abstract_call_gf_by_type`](https://github.com/JuliaLang/julia/blob/tb/inference_params/base/inference.jl#L923-L925). Any reason this should always be saving its results?

Performance difference in loading `base/inference.jl` seems undetectable. Suggestions for better testing of type inference performance?

cc @JeffBezanson @vtjnash

JeffBezanson commented:
Good tests of inference performance are the sys.so build, and some of the longer-running test suites like subarray.

JeffBezanson commented on base/inference.jl:
```diff
@@ -2,14 +2,44 @@
+    cached::Bool
+
+    # parameters limiting potentially-infinite types
+    MAX_TYPEUNION_LEN::Int
```
I would just keep all of these as global constants unless you have a use case for changing them on some calls.

JeffBezanson commented:
I don't really like that this makes a whole bunch of simple pure functions depend on an extra parameter that they usually don't use.

JeffBezanson commented:
> However, using cached that way breaks abstract_call_gf_by_type. Any reason this should always be saving its results?

The only reason we currently pass `cached=false` is to look at a single function e.g. without inlining. Any other functions recursively inferred during that process can still be optimized and cached as normal. In fact it is important *not* to propagate the parameters passed to inference, or we can end up caching non-optimized versions of functions. For example if I want to see the IR for `f() = 1+1` without inlining, we will end up inferring `+(::Int,::Int)` in the process, so we might as well save the result. The cache also prevents recursive functions from being inferred infinitely.

maleadt commented:
> I don't really like that this makes a whole bunch of simple pure functions depend on an extra parameter that they usually don't use.

I agree, but I felt like keeping some `const` and others configurable was a half-assed job. I'll try to strike a better balance, IIRC most of the diff is caused by`MAX_TUPLETYPE_LEN` and `MAX_TYPE_DEPTH` anyway.


> The cache also prevents recursive functions from being inferred infinitely.

Right, didn't realize this, that explains the hang.


> In fact it is important _not_ to propagate the parameters passed to inference, or we can end up caching non-optimized versions of functions.

OK, so I've changed that behavior in this PR, propagating all parameters but not caching the result instead.

The reason I wanted to propagate parameters for an entire inference "cycle", is because of the hooks I'm testing in #18338. Replacing a call applies to every function in a cycle, not only the top one. By propagating all parameters, including `cached`, we avoid re-using or polluting anything in the cache (except that it breaks recursion handling). This also side-steps the issue of needing multiple caches.

maleadt committed:
Put inference parameters in a dedicated object.

This allows overriding parameters, while reducing global state
and simplifying the API.

maleadt commented:
Updated and rebased.

I would still like to change the caching behavior, where inferring with `cached=false` adds called functions to the cache in order to break recursion. I don't fully grok the interplay between inference & the method tables, so any suggestions on how to do this @JeffBezanson @vtjnash?

vtjnash commented on base/inference.jl:
```diff
@@ -3951,10 +3980,13 @@ function reindex_labels!(sv::InferenceState)
 
-function return_type(f::ANY, t::ANY)
+function return_type(f::ANY, t::ANY, params::InferenceParams=InferenceParams())
+    # TODO: are all local calls to return_type processed by pure_eval_call?
```
no

vtjnash commented on base/inference.jl:
```diff
@@ -1590,7 +1618,7 @@ function typeinf_edge(method::Method, atypes::ANY, sparams::SimpleVector, caller
         end
     end
-    frame = typeinf_frame(code, true, true, caller)
+    frame = typeinf_frame(code, caller, InferenceParams(caller.params; cached=true))
```
this is a hot path, it would be nice to avoid the kwargs dispatch and allocation here

maleadt replied:
OK, removed.

vtjnash commented:
Can we leave out `cached` for now? I'll try to help you with making a selectable cache, but for now, I'm concerned it will just make #17057 trickier for me, since we're already overloading its meaning somewhat.

maleadt commented:
Yeah sure, I almost did already but left it in not knowing what a better solution would look like.
Dropping `cached` also obviates the need for the ugly kwarg copy ctor.

I can also maintain this PR until after #17057, if you like.

vtjnash commented:
the rest of it shouldn't conflict (other than superficially, and I already know I'll have to deal with rebasing other stuff anyways)

maleadt commented:
I've also left out `optimize`, because without proper caching (or copying-and-modifying the InferenceParams when passing down) a nonoptimized method could end up in the cache.

tkelman commented on base/inference.jl:
```diff
@@ -5,11 +5,25 @@ import Core: _apply, svec, apply_type, Builtin, IntrinsicFunction, MethodInstanc
+    # reasonable defaults
+    InferenceParams(;inlining::Bool=inlining_enabled(),
+                    tupletype_len::Int=15, tuple_depth::Int=16,
+                    tuple_splat::Int=4, union_splitting::Int=4) =
```
this is swapping the defaults for depth and splat

maleadt replied:
That's bad. Nicely spotted, fixed.

JeffBezanson commented:
Is there a particular reason those 4 globals were moved into the struct, and the other 2 were kept global?

maleadt commented:
I initially put them all in the struct, but that caused too much churn (as you mentioned above, making a whole bunch of simple pure functions depend on an extra parameter). So I took the ones causing the largest diff out.

andyferris commented:
Out of curiosity, when we say "configurable" is e.g. `MAX_TUPLE_SPLAT` something I can specify on startup, when building a system image, or at the REPL?

maleadt commented:
@andyferris: no, apart from changing the default value in `inference.jl`, or when calling into inference yourself (eg. [this demo](https://github.com/JuliaLang/julia/blob/02eebd316a4c44f035577ac4137bf3e96a24fbcc/demo.jl#L44-L47)).

andyferris commented:
Ok - thanks @maleadt
